halo/evmengine/keeper/abci_internal_test.go

```diff
@@ -0,0 +1,544 @@
+				},
+				wantErr: true,
+			},
+		}
+		for _, tt := range tests {
+			t.Run(tt.name, func(t *testing.T) {
+				t.Parallel()
+				ctx, storeService := setupCtxStore(t)
+				cdc := getCodec(t)
+				txConfig := authtx.NewTxConfig(cdc, nil)
+				ap := MockAddressProvider{}
+
+				k := NewKeeper(cdc, storeService, &tt.mockEngine, txConfig, ap)
+				_, err := k.PrepareProposal(ctx, tt.req)
+				if (err != nil) != tt.wantErr {
+					t.Errorf("PrepareProposal() error = %v, wantErr %v", err, tt.wantErr)
+					return
+				}
+			})
+		}
+	})
+
+	t.Run("build non optimistic", func(t *testing.T) {
+		t.Parallel()
+		ctx, storeService := setupCtxStore(t)
+		cdc := getCodec(t)
+		txConfig := authtx.NewTxConfig(cdc, nil)
+
+		me, err := ethclient.NewEngineMock()
+		require.NoError(t, err)
+		mockEngine := MockEngineAPI{
+			Mock:   me,
+			fuzzer: ethclient.NewFuzzer(time.Now().Truncate(time.Hour * 24).Unix()),
+		}
+		ap := MockAddressProvider{}
+		keeper := NewKeeper(cdc, storeService, &mockEngine, txConfig, ap)
+
+		ts := time.Now()
+		latestHeight, err := mockEngine.BlockNumber(ctx)
+		require.NoError(t, err)
+		latestBlock, err := mockEngine.BlockByNumber(ctx, big.NewInt(int64(latestHeight)))
+		require.NoError(t, err)
+
+		mockEngine.pushPayload(t, ctx, ap.LocalAddress(), latestBlock.Hash(), ts)
+		nextBlock, blockPayload := mockEngine.nextBlock(t, latestHeight+1, uint64(ts.Unix()), latestBlock.Hash(), ap.LocalAddress())
+		_, err = mockEngine.Mock.NewPayloadV2(ctx, blockPayload)
+		require.NoError(t, err)
+		payloadID := mockEngine.pushPayload(t, ctx, ap.LocalAddress(), nextBlock.Hash(), ts)
+
+		req := &abci.RequestPrepareProposal{
+			Txs:    nil,
+			Height: int64(2),
+			Time:   time.Now(),
+		}
+
+		keeper.mutablePayload.Height = 2
+		keeper.mutablePayload.UpdatedAt = time.Now()
+		keeper.mutablePayload.ID = payloadID
+
+		resp, err := keeper.PrepareProposal(ctx, req)
+		require.NoError(t, err)
+		require.NotNil(t, resp)
+
+		tx, err := txConfig.TxDecoder()(resp.Txs[0])
+		require.NoError(t, err)
+
+		for _, msg := range tx.GetMsgs() {
+			if _, ok := msg.(*etypes.MsgExecutionPayload); ok {
+				assertExecutablePayload(t, msg, ts.Unix(), nextBlock.Hash(), ap, uint64(req.Height))
+			}
+		}
+	})
+
+	t.Run("build optimistic", func(t *testing.T) {
+		t.Parallel()
+		ctx, storeService := setupCtxStore(t)
+		cdc := getCodec(t)
+		txConfig := authtx.NewTxConfig(cdc, nil)
+
+		me, err := ethclient.NewEngineMock()
+		require.NoError(t, err)
+		mockEngine := MockEngineAPI{
+			Mock:   me,
+			fuzzer: ethclient.NewFuzzer(time.Now().Truncate(time.Hour * 24).Unix()),
+		}
+		ap := MockAddressProvider{}
+		keeper := NewKeeper(cdc, storeService, &mockEngine, txConfig, ap)
+
+		ts := time.Now()
+		latestHeight, err := mockEngine.BlockNumber(ctx)
+		require.NoError(t, err)
+		latestBlock, err := mockEngine.BlockByNumber(ctx, big.NewInt(int64(latestHeight)))
+		require.NoError(t, err)
+
+		mockEngine.pushPayload(t, ctx, ap.LocalAddress(), latestBlock.Hash(), ts)
```

Comment:
add much more comments, one for every step, it really isn't very clear what is happening here.